### DISCUSSION TITLE
Getting Shared Memory to work between sandboxed application and non-sandboxed application?

### DISCUSSION THREAD

**Discussion by myanowitz** [netblue30/firejail#4751] [Q&A]
It works perfectly on Ubuntu 18.04 Linux, but I am having issues running similar code on a aarch64 Yocto Linux.
Shared Memory does not appear to be working between inside Firejail and outside Firejail. Not sure why. It works fine when both apps are outside of Firejail.
Can someone clue me in on how to debug this?

Adding these does not help:

```sh
firejail --noprofile --keep-dev-shm --noblacklist=/dev/shm sh
```

Adding these causes an error:

```sh
firejail --noprofile --seccomp.keep=shmat,shmdt,shmctl,shmget,sem_open,sem_close,sem_unlink sh
```

yields:

```
Parent pid 93078, child pid 93079
The new log directory is /proc/93079/root/var/log
Child process initialized in 18.52 ms

Parent pid 93078, child pid 93079
The new log directory is /proc/93079/root/var/log
Child process initialized in 18.52 ms

Parent pid 93078, child pid 93079
The new log directory is /proc/93079/root/var/log
Child process initialized in 18.52 ms

Parent is shutting down, bye...
```

I am hoping the reason has nothing to do with AppArmor.

If I remove the --noprofile, I get this:

```console
root@aptiv-ep2000:~# firejail --noprofile --seccomp.keep=shmat,shmdt,shmctl,shmget,sem_open,sem_close,sem_unlink sh
Parent pid 93078, child pid 93079
The new log directory is /proc/93079/root/var/log
Child process initialized in 18.52 ms

Parent is shutting down, bye...
```

```console
root@aptiv-ep2000:~# firejail --keep-dev-shm --noblacklist=/dev/shm --seccomp.keep=shmat,shmdt,shmctl,shmget,sem_open,sem_close,sem_unlink sh
Reading profile /etc/firejail/server.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-passwdmgr.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/disable-write-mnt.inc
Reading profile /etc/firejail/disable-xdg.inc

** Note: you can use --noprofile to disable server.profile **

Parent pid 93597, child pid 93598
The new log directory is /proc/93598/root/var/log
Warning: cannot open destination file /home/root/.bashrc, file not copied
Warning: cannot copy /etc/skel/.bashrc
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Warning: Cannot confine the application using AppArmor.
Maybe firejail-default AppArmor profile is not loaded into the kernel.
As root, run "aa-enforce firejail-default" to load it.
Child process initialized in 102.01 ms

Parent is shutting down, bye...
```

I simply want shared memory to work between an application in firejail and one outside firejail.
How should I accomplish this?

---

Edit by @kmk3: Formatting/grammar.

**Comment by rusty-snake** [comment]
Does it work with the `noprofile.profile`?

> --seccomp.keep=shmat,shmdt,shmctl,shmget,sem_open,sem_close,sem_unlink

You need to allow syscalls like `open(at)`,`(p)write(v|v2|64)`,`(p)read(v|v2|64)`,`close`,`dup(2|3)`,`lseek`,`(f)access(at)`,`(f)stat(at)`,...

**Reply by myanowitz** [reply]
Thanks for the response.
But, no, using --noprofile does not work here.

Adding the additional system calls did not work, maybe I need to add more:
```
~/bin# firejail --noprofile --seccomp.keep=shmat,shmdt,shmctl,shmget,sem_open,sem_close,sem_unlink,open,openat,write,pwrite,writev,write64,pwrite,read,pread,read64,close,dup,dup2,dup3,lseek,access,faccess,stat,fstat --keep-dev-shm sh
Parent pid 16053, child pid 16054
The new log directory is /proc/16054/root/var/log
Child process initialized in 18.67 ms

Parent is shutting down, bye...
```

If I do this, at-least it stays in the firejail shell, but I still don't have access to shared memory:
```
~/bin# firejail --noprofile --seccomp.drop=NONE --keep-dev-shm --read-write=/dev/shm --noblacklist=/dev/shm sh
Seccomp list in: NONE check list: @default-keep,
Parent pid 21902, child pid 21903
The new log directory is /proc/21903/root/var/log
Seccomp list in: NONE check list: @default-keep,
Child process initialized in 21.82 ms
sh-5.0# 

```

**Reply by rusty-snake** [reply]
>  But, no, using --noprofile does not work here.

noprofile**.profile** (download it from master)

> Adding the additional system calls did not work, maybe I need to add more:

Yes, a lot more.

> --seccomp.drop=NONE

NONE isn't a valid syscall.

> --noblacklist=/dev/shm

never blacklisted.

**Comment by myanowitz** [comment]
Thanks for the suggestions, but still not working for me:

I tried your _noprofile.profile_
I tried with and without adding
_ignore seccomp_
I had to copy seccomp.debug to seccomp.debug32

But I am still getting:
```
root@aptiv-ep2000:~/bin# firejail --profile=noprofile  --debug --keep-dev-shm sh
Reading profile /etc/firejail/noprofile.profile
Autoselecting /bin/sh as shell
Building quoted command line: 'sh' 
Command name #sh#
DISPLAY is not set
Enabling IPC namespace
Using the local network stack
Initializing child process
Parent pid 67164, child pid 67165
The new log directory is /proc/67165/root/var/log
Host network configured
PID namespace installed
Mounting tmpfs on /run/firejail/mnt directory
Creating empty /run/firejail/mnt/seccomp directory
Creating empty /run/firejail/mnt/seccomp/seccomp.protocol file
Creating empty /run/firejail/mnt/seccomp/seccomp.postexec file
Creating empty /run/firejail/mnt/seccomp/seccomp.postexec32 file
Mounting /proc filesystem representing the PID namespace
Basic read-only filesystem:
Mounting read-only /usr
506 465 0:21 /usr /usr ro,relatime master:1 - overlay overlayfs-root rw,lowerdir=/mnt/lower,upperdir=/mnt/usr/rw_psdk7/upper,workdir=/mnt/usr/rw_psdk7/work
mountid=506 fsname=/usr dir=/usr fstype=overlay
Mounting read-only /bin
507 465 0:21 /bin /bin ro,relatime master:1 - overlay overlayfs-root rw,lowerdir=/mnt/lower,upperdir=/mnt/usr/rw_psdk7/upper,workdir=/mnt/usr/rw_psdk7/work
mountid=507 fsname=/bin dir=/bin fstype=overlay
Mounting read-only /sbin
508 465 0:21 /sbin /sbin ro,relatime master:1 - overlay overlayfs-root rw,lowerdir=/mnt/lower,upperdir=/mnt/usr/rw_psdk7/upper,workdir=/mnt/usr/rw_psdk7/work
mountid=508 fsname=/sbin dir=/sbin fstype=overlay
Mounting read-only /lib
509 465 0:21 /lib /lib ro,relatime master:1 - overlay overlayfs-root rw,lowerdir=/mnt/lower,upperdir=/mnt/usr/rw_psdk7/upper,workdir=/mnt/usr/rw_psdk7/work
mountid=509 fsname=/lib dir=/lib fstype=overlay
Mounting tmpfs on /var/lock
Create the new utmp file
Mount the new utmp file
blacklist /run/firejail/dbus
Mounting read-only /proc/sys
Remounting /sys directory
Disable /sys/firmware
Disable /sys/power
Disable /sys/kernel/debug
Disable /sys/kernel/vmcoreinfo
Disable /proc/sys/kernel/core_pattern
Disable /proc/sys/kernel/modprobe
Disable /proc/sysrq-trigger
Disable /proc/sys/vm/panic_on_oom
Disable /proc/irq
Disable /proc/bus
Disable /proc/sched_debug
Disable /proc/timer_list
Disable /proc/kallsyms
Disable /dev/port
Not blacklist /sys/fs
Not blacklist /sys/module
Current directory: /home/root/bin
Mounting read-only /run/firejail/mnt/seccomp
482 503 0:47 /seccomp /run/firejail/mnt/seccomp ro,nosuid - tmpfs tmpfs rw,mode=755
mountid=482 fsname=/seccomp dir=/run/firejail/mnt/seccomp fstype=tmpfs
Seccomp directory:
ls /run/firejail/mnt/seccomp
drwxr-xr-x root     root             120 .
drwxr-xr-x root     root             120 ..
-rw-r--r-- root     root              40 seccomp
-rw-r--r-- root     root              40 seccomp.32
-rw-r--r-- root     root               0 seccomp.postexec
-rw-r--r-- root     root               0 seccomp.postexec32
No active seccomp files
Drop CAP_DAC_OVERRIDE
Drop CAP_DAC_READ_SEARCH
Drop privileges: pid 1, uid 0, gid 0, nogroups 0
No supplementary groups
Starting application
LD_PRELOAD=(null)
Running 'sh'  command through /bin/sh
execvp argument 0: /bin/sh
execvp argument 1: -c
execvp argument 2: 'sh' 
Child process initialized in 22.47 ms
sh-5.0# monitoring pid 2

./ShmConsumer 
Enter the shm id to attach to: 
14
Unable to attach to shared memory space
Press a key to finish

```

I tried the --seccomp.drop=NONE
as I was hoping that would KEEP all system calls except something called NONE wheter or not NONE exists.

**Comment by kranthikud** [comment]
> t works perfectly on Ubuntu 18.04 Linux, but I am having issues running similar code on a aarch64 Yocto Linux. Shared Memory does not appear to be working between inside Firejail and outside Firejail. Not sure why. It works fine when both apps are outside of Firejail. Can someone clue me in on how to debug this. Adding these do not help: firejail --noprofile **_--keep-dev-shm --noblacklist=/dev/shm_** sh
> 
> Adding these cause an error: firejail --noprofile **_--seccomp.keep=shmat,shmdt,shmctl,shmget,sem_open,sem_close,sem_unlink_** sh yields:
> 
> Parent pid 93078, child pid 93079 The new log directory is /proc/93079/root/var/log Child process initialized in 18.52 ms
> 
> Parent pid 93078, child pid 93079 The new log directory is /proc/93079/root/var/log Child process initialized in 18.52 ms
> 
> Parent pid 93078, child pid 93079 The new log directory is /proc/93079/root/var/log Child process initialized in 18.52 ms
> 
> **Parent is shutting down, bye...**
> 
> I am hoping the reason has nothing to do with AppArmor, (if I remove the --noprofile, I get this):
> 
> root@aptiv-ep2000:~# firejail --noprofile --seccomp.keep=shmat,shmdt,shmctl,shmget,sem_open,sem_close,sem_unlink sh Parent pid 93078, child pid 93079 The new log directory is /proc/93079/root/var/log Child process initialized in 18.52 ms
> 
> Parent is shutting down, bye... root@aptiv-ep2000:~# firejail --keep-dev-shm --noblacklist=/dev/shm --seccomp.keep=shmat,shmdt,shmctl,shmget,sem_open,sem_close,sem_unlink sh Reading profile /etc/firejail/server.profile Reading profile /etc/firejail/disable-common.inc Reading profile /etc/firejail/disable-passwdmgr.inc Reading profile /etc/firejail/disable-programs.inc Reading profile /etc/firejail/disable-write-mnt.inc Reading profile /etc/firejail/disable-xdg.inc
> 
> ** Note: you can use --noprofile to disable server.profile **
> 
> Parent pid 93597, child pid 93598 The new log directory is /proc/93598/root/var/log Warning: cannot open destination file /home/root/.bashrc, file not copied Warning: cannot copy /etc/skel/.bashrc Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set. Warning: Cannot confine the application using AppArmor. Maybe firejail-default AppArmor profile is not loaded into the kernel. As root, run "aa-enforce firejail-default" to load it. Child process initialized in 102.01 ms
> 
> Parent is shutting down, bye...
> 
> I simply want shared memory to work between an application in firejail and one outside firejail. How should I accomplish this.

@myanowitz, facing similar issue. Do you got any solution make this to work? Thanks.

**Reply by rusty-snake** [reply]
>  --seccomp.keep=shmat,shmdt,shmctl,shmget,sem_open,sem_close,sem_unlink 

Denying almost all syscalls including basic IO calls like `read` and `write` will hardly work for any program.

**Reply by kranthikud** [reply]
Nope. This is not working. I think firejail has some static configurations to run any application. The user does not have full control actually.

**Comment by myanowitz** [comment]
I’ll get back to you on this next week.
I made little further progress, and have been working on other projects.
Someone else on my team was working on this, and probably has a solution
but he is on vacation this week. I will ask next week.

Michael J Yanowitz
Software Engineer
[aptiv]<http://www.aptiv.com/>
Connected Services
Office: +1 (860)371-0265  | Mobile: +1 (734)450-9309

999 Republic Dr., Suite 100 | Allen Park, MI, 48101

---

Edit by @kmk3: Remove email top-posting noise.

**Reply by kmk3** [reply]
(Offtopic)

@myanowitz [on Jan 5](https://github.com/netblue30/firejail/discussions/4751#discussioncomment-4600556):

> I’ll get back to you on this next week. I made little further progress, and
> have been working on other projects. Someone else on my team was working on
> this, and probably has a solution but he is on vacation this week. I will ask
> next week.

Out of curiosity, are you working on firejail as part of a larger project?

I only remember seeing SailfishOS using it (as they contributed some of their
patches); would be interesting to see firejail being used in more places.

**Comment by myanowitz** [comment]
Yes, I am working on another large  project, (which I promised I wouldn’t give any details),
where we run some services sandboxed so they can be isolated. We were/are planning to use
Firejail for that purpose. At this point it is still being developed.

Michael J Yanowitz
Software Engineer
[aptiv]<http://www.aptiv.com/>
Connected Services
Office: +1 (860)371-0265  | Mobile: +1 (734)450-9309

999 Republic Dr., Suite 100 | Allen Park, MI, 48101

---

Edit by @kmk3: Remove email top-posting noise.

**Reply by kmk3** [reply]
@myanowitz [on Jan 9](https://github.com/netblue30/firejail/discussions/4751#discussioncomment-4633950):

> Yes, I am working on another large project, (which I promised I wouldn’t give
> any details), where we run some services sandboxed so they can be isolated.
> We were/are planning to use Firejail for that purpose. At this point it is
> still being developed.

I see; nice to see it being considered as a tool for isolation either way.

Feel free to share more information if/when you can.

Would be interesting to know how it compares to the alternatives in that case
and what could be improved.

**Comment by kmk3** [comment]
@myanowitz [on Dec 8, 2021](https://github.com/netblue30/firejail/discussions/4751#discussion-3729091):

> It works perfectly on Ubuntu 18.04 Linux, but I am having issues running
> similar code on a aarch64 Yocto Linux. Shared Memory does not appear to be
> working between inside Firejail and outside Firejail.

What is the version of Yocto being used?

What is the output of `uname -srmo` in Yocto?

What is the output of `firejail --version` in Yocto?

If I understand it correctly, /dev/shm is basically the same thing as /tmp
except that it is guaranteed to be mounted as tmpfs.

On my system I can create a file in /dev/shm outside of the sandbox and read it
inside of the sandbox:

```console
$ echo 123 >/dev/shm/123
$ cat /dev/shm/123
123
$ firejail --quiet --profile=noprofile cat /dev/shm/123
123
```

Environment: firejail-git (15011a6d8) on Artix Linux.

What happens when trying to do that in Yocto?

Does the issue still happen with the current version of firejail from master?

Does Yocto do anything special with /dev/shm?

To get more debugging information, what is the output of the following
commands?

Without firejail:

```sh
ls -l /dev/shm
stat /dev/shm
mount | grep /dev/shm
```

With firejail:

```
firejail --quiet --profile=noprofile ls -l /dev/shm
firejail --quiet --profile=noprofile stat /dev/shm
firejail --quiet --profile=noprofile mount | grep /dev/shm
```

Note: On my system there is no difference in the above outputs when comparing
with/without firejail.

**Reply by kranthikud** [reply]
**firejail version 0.9.57**

The above shared memory related commands are working fine in firejail.
 
The firejail command with argument of --seccomp.keep is always killing the parent process.

```sh
firejail --noprofile --seccomp.keep=shmat,shmdt,shmctl,shmget --debug ssh
```

The following are the debug messages:

```
seccomp filter configured
Drop privileges: pid 1, uid 0, gid 0, nogroups 0
No supplementary groups
starting application
LD_PRELOAD=(null)
Running 'ssh'  command through /bin/sh
execvp argument 0: /bin/sh
execvp argument 1: -c
execvp argument 2: 'ssh'
Child process initialized in 18.20 ms
Installing /run/firejail/mnt/seccomp seccomp filter
monitoring pid 4

Sandbox monitor: waitpid 4 retval 4 status 159

Parent is shutting down, bye...
```

**Reply by kmk3** [reply]
@kranthikud [on Jan 10](https://github.com/netblue30/firejail/discussions/4751#discussioncomment-4640883):

> **firejail version 0.9.57**

On what distribution/version?

Note that 0.9.56~0.9.58.2 are about 4 years old and are not supported by us:

* <https://github.com/netblue30/firejail/blob/master/SECURITY.md>

Also, such old versions may be affected by multiple security issues that were
fixed over the years, such as:

* [CVE-2019-12589](https://github.com/advisories/GHSA-q9v9-56wp-f76g) (fixed in 0.9.60)

See also <https://github.com/netblue30/firejail#installing>.

> The above shared memory related commands are working fine in firejail.
> 
> The firejail command with argument of --seccomp.keep is always killing the
> parent process.
> 
> ```shell
> firejail --noprofile --seccomp.keep=shmat,shmdt,shmctl,shmget --debug ssh
> ```

That's likely because ssh uses way more syscalls than just those 4.

To debug the blocked syscalls, run `dmesg -H` and see the audit logs.  Example:

```
[...] audit: [...] comm="cat" exe="/usr/bin/cat" sig=0 arch=c000003e syscall=21 compat=0 [...]
[...] audit: [...] comm="cat" exe="/usr/bin/cat" sig=0 arch=c000003e syscall=1 compat=0 [...]
```

The names of the blocked syscalls can be looked up with `--debug-syscalls` /
`--debug-syscalls32` (depending on the value of `arch=` above):

```console
$ firejail --debug-syscalls | grep -E '^(1|21)\s'
1       - write
21      - access
```

When trying to use `seccomp.keep` with the `cat` example, all of the following
syscalls appeared in the audit log, and allowing them in `seccomp.keep` made
the command work:

```console
$ firejail --quiet --profile=noprofile \
  --seccomp.keep=access,arch_prctl,brk,close,exit,exit_group,fadvise64,getrandom,mmap,mprotect,munmap,open,openat,pread64,prlimit64,read,rseq,newfstatat,set_robust_list,set_tid_address,shmget,write,writev \
  cat /dev/shm/123
123
```

Not necessarily all of them are strictly needed (and I suppose that some may
only be used when others fail), but the above works.  See also syscall groups,
such as `@default-keep`.  Note that this is not very documented in the man
pages; use something like `git grep '@default-keep'` instead.

By the way, as of firejail 0.9.72-rc1 commands are executed directly by default
(without going through `sh -c`), which should help reduce the number of needed
syscalls (depending on the program being executed).

**Reply by kmk3** [reply]
(Offtopic)

@kranthikud

Please see the following links for how to format code blocks in markdown:

* <https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks>
* <https://github.github.com/gfm/#fenced-code-blocks>

**Reply by kranthikud** [reply]
@kmk3 , thanks. I will try with latest version of firejail and come back. But I have one basic question with the firejail.
Even though I have run the firejail with `--noprofile `option, why is it still restricting some of the system calls? Does firejail always starts with some static options that user does not have any control?

**Reply by rusty-snake** [reply]
> why is it still restricting some of the system calls?

If you run with `--noprofile` firejail will not restrict any syscall (unless you provide additional arguments like `--seccomp`). However, other options may cause Kernel security checks to fail or resources to be unavailable.

>  Does firejail always starts with some static options that user does not have any control?

Correct, even with `--noprofile` there are still some minimal restrictions.
1. Options you need to allow explicit. To allow them all run `firejail --profile=noprofile` ([noprofile.profile](https://github.com/netblue30/firejail/blob/master/etc/profile-m-z/noprofile.profile)) <sub>firejail >= 0.9.68</sub>
2. Options you do not have any control about. Most notably the hardcoded blacklist, unsharing of the pid-namespace and NNP if configured.
